ocaml/xapi/cert_distrib.ml

```diff
@@ -0,0 +1,247 @@
+module Unixext = Xapi_stdext_unix.Unixext
+
+module D = Debug.Make (struct let name = "cert_distrib" end)
+
+module XenAPI = Client.Client
+
+type existing_cert_strategy = Erase | Merge [@@deriving sexp]
+
+type 'a remote_call =
+  API.ref_host -> (Rpc.call -> Rpc.response) -> API.ref_session -> 'a
+
+module WireProtocol = struct
+  module Sexp = Sexplib.Sexp
+  open Sexplib0.Sexp_conv
+
+  type cert_blob = string [@@deriving sexp]
+
+  type cert = {uuid: string; blob: cert_blob} [@@deriving sexp]
+
+  type payload =
+    | Collect
+    | Write of (existing_cert_strategy * cert list)
+    | GenBundle
+  [@@deriving sexp]
+
+  type result = CollectResult of cert_blob | WriteResult | GenBundleResult
+  [@@deriving sexp]
+
+  let string_of_payload x = x |> sexp_of_payload |> Sexp.to_string
+
+  let payload_of_string x =
+    try Some (x |> Sexp.of_string |> payload_of_sexp) with _ -> None
```

Comment:
Well the alternative is to just have separate messages in the datamodel for each step in the distribution procedure. This is what we do for Pool secret rotation (for example: https://github.com/xapi-project/xen-api/blob/6eedefc121d3015d73f2978d9c8871fee551fcbe/ocaml/xapi/xapi_host.ml#L2421)

I mentioned my motivation for defining my own protocol below:
```
(* eventually the remote calls should probably become API calls in the datamodel
 * but they remain here for quick development *)
```

Personally I prefer the current implementation because _pretty much_ everything to do with cert distribution can be kept in just this file, and we don't have to expose internal details in the datamodel. However I also appreciate the argument for putting all of these calls in datamodel, as that gives us explicit versioning

Reply:
Don't worry about this, I'm just rambling because I haven't found a nice way to get rid of the `base` dependency ;)

Reply:
There is a `sexplib0` package from JaneStreet with zero dependencies. Would that work?

Reply:
Unfortunately, no. sexplib0 is unable to parse a string to generate a structured s-expression.